src/java/org/apache/cassandra/index/sai/plan/Expression.java

```diff
@@ -150,7 +150,9 @@ public boolean isContains()
     private float searchRadiusMeters = 0;
     private float searchRadiusDegreesSquared = 0;
     public int topK;
-    public boolean upperInclusive, lowerInclusive;
+    // These variables are only meant to be used for final validation of the range search. They are not
+    // meant to be used when searching the index. See the 'add' method below for additional explanation.
```

Comment:
If they are not used when searching the index, then why  do we use them in `estimateMatchingRowsCount`? 
The estimate should match index search as closely as possible.

Reply:
Look closely, we don't use them there after one of my commits.

> The estimate should match index search as closely as possible.

Agreed, and that was the problem. If you look below in this class at the `add` method, you'll notice that we shadow the variable names with local variables of the same names. Those then get different treatment for types that "round" or truncate the values in the index.

These booleans here are not used for the actual index search, but are instead used for index validation.

And as I write this, I just noticed that this means we probably have an issue for NOT on `decimal` and `varint` because those are truncated and we don't know if there is more than one value for a given entry in the index.

Reply:
Here is a failing test that shows the broken NOT queries:

```java
    @Test
    public void testLossyDecimalQueries() throws Throwable
    {
        createTable("CREATE TABLE %s (pk text PRIMARY KEY, x decimal)");
        createIndex("CREATE CUSTOM INDEX ON %s(x) USING 'StorageAttachedIndex'");

        // We truncate decimals to 24 bytes in the numeric index. These numbers are chosen to ensure
        // that we get the expected results.
        var a = "1111111111111111111111111111111111111111111111111111111111";
        var b = "1111111111111111111111111111111111111111111111111111111112";
        var c = "1111111111111111111111111111111111111111111111111111111113";

        execute("INSERT INTO %s (pk, x) VALUES ('a', " + a + ')');
        execute("INSERT INTO %s (pk, x) VALUES ('b', " + b + ')');
        execute("INSERT INTO %s (pk, x) VALUES ('c', " + c + ')');

        beforeAndAfterFlush(() -> {
            // Test two kinds of NOT queries
            assertRows(execute("SELECT pk FROM %s WHERE x NOT IN (" + b + ')'), row("a"), row("c"));
            assertRows(execute("SELECT pk FROM %s WHERE x != " + b), row("a"), row("c"));
        });
    }
```

Reply:
Addressed, tests added.